src/Preflight/Preflight.php

```diff
@@ -249,7 +249,7 @@ public function preflight($argv)
         // Extend configuration and alias files to include files in
         // target site.
         $root = $this->findSelectedSite();
-        $this->configLocator->addSitewideConfig($root);
+        $this->configLocator->addDrupalConfig($root);
```

Comment:
In Drush8, we don't load any site specific config until the Site phase of the bootstrap. I'd like to keep consistent with that in Drush9. Since this change is in Preflight (i.e. happens before bootstrap), I dont think its properly placed.

Reply:
I think that moving all config processing to preflight is an improvement. When config was php, delaying loading site-specific config until site bootstrap has the benefit of allowing folks to use Drupal APIs in their config code. With yaml config, I'm not sure that there is a benefit to postponing config loading. Loading config in preflight might allow folks to use `--uri=foo` and set URI to `https://foo.example.com` in their drush.yml file, to give one example.

Reply:
That is exactly my use case.

Reply:
I think we are working around the problem that config is static after its loaded. It would be preferable IMO if uri were discovered during SITE phase and then used for link building and such during command execution.

We already do ROOT discovery during Preflight. This PR effectively moves SITE discovery there too. It begs the question of how much do we need the progressive Bootstrap that we do. Maybe NONE and FULL is all we need? We don't need to figure this out now but it would be good to get some clarity on future directions.

Reply:
I think that site selection is already happening in preflight. Note the comment removed in this PR: "// not implemented yet (multisite)" (see my comment https://github.com/drush-ops/drush/pull/3278#pullrequestreview-86221163). I think this PR is essentially implementing a TODO. I still don't see any practical benefit to postponing uri discovery to the SITE phase, and as previously mentioned, there are some disadvantages to doing that.

Regarding progressive bootstrap, for command discovery we only need NONE (built-in and global commands) and FULL (module commands).  Built-in and global commands can also declare the bootstrap level they want, but it is not possible for them to declare their bootstrap level, because we already have to bootstrap to FULL in order to find them.

Reply:
I could be wrong, but I think this moves the inclusion of sites.php to Preflight from SITE phase. That might be safe, but I'm not sure what people are doing there.

Reply:
Actually, I think that prior to this PR, `sites.php` was never included by Drush9.

Yes it is now included during preflight if a uri is specified. Given that `sites.php` should only ever contain a static PHP array, this does not seem problematic. 

From a conceptual standpoint, `sites.php` really contains data that is above the "site" context. It contains data about _all_ sites belonging to the application, which is used to determine the appropriate site context.

Reply:
I agree this is pretty safe and and agree with your conceptual point.

FYI, Drush is relying on Drupal to include sites.php. This happens during the SITE phase I have been mentioning. See https://github.com/drush-ops/drush/blob/3fc8e7156ea08081eadedc4f61fe83315fa1f883/src/Boot/DrupalBoot8.php#L73-L83. I dont love duplicating that logic in Drush, but I may tolerate it.

Reply:
Ah, okay, I see now.

We really must include drush.yml files from the sites directory during preflight. In order to do this, we need to identify the sites directory, and in order to identify the sites directory (the same way Drupal does) we need to include sites.php.

I'm not fond of putting `lookupSiteDirFromUri` in StringUtil. I would also prefer to keep Drupal-specific code out of preflight. I'd also like to continue to rely on `DrupalKernel::findSitePath` to process sites.php.

Some of these desires are mutually exclusive. I'm not sure what the best compromise is. To whit:

1. If we do not include drush.yml in the preflight (or at least at the beginning of the bootstrap site phase), then we limit our ability to use `--uri=dir` and set a better uri in drush.yml.
1. If we do not include sites.php at all, then we limit our ability to use `--uri=https://example.com` and still find the right sites directory, and finding the right sites directory is necessary to include drush.yml soon enough to improve the uri.

I think it would be worthwhile to try harder to use `DrupalKernel::findSitePath` to find the site path and load drush.yml. Can this happen at the beginning of the site bootstrap phase, before we've done the site bootstrap, so that we can then set the uri and continue with the site bootstrap? Or do we need the uri to make the request object needed to call `DrupalKernel::findSitePath`?  Maybe the load order of drush.yml during the site bootstrap phase needs to vary depending on whether uri is a directory name or contains a protocol? Having a conditional order of events in bootstrap isn't great either, but I'd strongly prefer to keep the Drupal-specific steps in bootstrap, if that's possible. Breaking that rule makes the Drush preflight / bootstrap harder to understand.

Reply:
The code I linked to already runs during SITE VALIDATE. We could certainly add a call to \Drush\Config\ConfigLocator::addConfigPaths during SITE phase. I'm not sure how to get access to the ConfigLocator instance though. Also, this would not satisfy the desire to set uri from a site specific config file. I honestly doubt that we can/should support that right now.

Reply:
By the current design, the config locator is not available during bootstrap.

Does it not currently work to set `$options['uri']` from a site-specific drushrc.php file in Drush 8 with Drupal 8?

Reply:
I dont know. If so, we just accept that as a regression with known workarounds (use an alias, use site-set, specify uri on the request, ...).

Reply:
Yeah I agree. While I'd like to see this feature work, I think it's enough of an edge case that I'd like the solution to be clean (keep Drupal-specific code in the bootstrap phase).

Reply:
I have found one more workaround. If you set the environment variable `DRUSH_OPTIONS_URI=http://example.com` then your custom uri will be used. No need to fuss with Drush config or aliases.

So, how to set that env variable? If you have many sites on same host, or don't control the server, this can be challenging. 
1. Anyone used [direnv](https://github.com/direnv/direnv)?
1. I'm using PHPEnv successfully on Drupal projects but at this time there is no way to load custom code early enough in a Drush request for 'uri' to take effect. That happens at end of Runtime https://github.com/drush-ops/drush/blob/master/src/Runtime/Runtime.php#L88-L93

I will now writeup docs for #3009 as I now understand that system.

Reply:
As a follow-up, the drupal-project will soon support .env files and documents how to DRUSH_OPTIONS_URI so that site specific works. See https://github.com/drupal-composer/drupal-project/pull/351. You can implement that in your composer project even before this PR is merged.